### DISCUSSION TITLE
Calling a python function of a user-provided script repeatedly in rust

### DISCUSSION THREAD

**Discussion by tajpulo** [RustPython/RustPython#5340] [Q&A]
I want to generate a static executable written in rust. When it is run, a user can provide a python script to customize the behavior of this rust program. Specifically, I want to run a hook at specific points in my rust program.

More specifically, rust traverses the file system, calls a user function per file basename and the user function can tell the updated basename.

# Python user script

```python
def rename(original_basename):
    if original_basename == "index.htm":
        return "index.html"
    return original_basename
```

# Cargo.toml (for the record)

```toml
[package]
name = "pyuserscript"
version = "0.1.0"
edition = "2021"

[dependencies]
rustpython = "0.3.1"
rustpython-vm = "0.3.1"
walkdir = "2.5.0"
```

# rust proof-of-concept

```rust
//! Application to run python userscript per filepath

use rustpython_vm as vm;
use rustpython;

use vm::builtins::{PyCode, PyStr};
use vm::compiler::CompileError;
use vm::{PyPayload, PyRef};
use rustpython_vm::compiler::core::Mode;

use walkdir::WalkDir;

use std::fs;
use std::path;

fn evaluate_filepath(interp: &vm::Interpreter, filepath: &path::Path, userscript_filepath: &path::Path) -> Result<(), String> {

    let filepath_basename = match filepath.file_name() {
        Some(b) => b,
        None => return Ok(()),
    };

    if let Err(err) = interp.enter(|rvm| {
        let source: String = fs::read_to_string(userscript_filepath).expect("cannot read python source file");

        let compiled: Result<PyRef<PyCode>, CompileError> = rvm.compile(&source, Mode::Exec, userscript_filepath.to_str().unwrap().to_string());
        match compiled {
            Ok(ref_code) => {
                let scope = rvm.new_scope_with_builtins();
                let object_ref = rvm.run_code_obj(ref_code, scope.clone())?;

                let fn1 = scope.globals.get_item("rename", rvm)?;
                let fn1_result = fn1.call((filepath_basename.to_str().unwrap().to_string(),), rvm)?;

                let ret_value: PyRef<PyStr> = match fn1_result.downcast() {
                    Ok(string_return_value) => string_return_value,
                    Err(_) => return Err(rvm.new_value_error(format!("function rename did not return a string"))),
                };
                println!("modified filepath = '{}'", ret_value);
                // TODO actually rename the file, not doing it so no copier accidentally runs it
            },
            Err(err) => {
                eprintln!("could not compile provided user script '{}'", filepath.display());
                eprintln!("error: {}", err);
                return Err(rvm.new_import_error(format!("failed to compile provided user script"), PyStr::from("userscript.py").into_ref(rvm.ctx.as_ref())));
            },
        }
        
        Ok(())
    }) {
        let e = err.clone();
        let errmsg = format!("cause={:?} context={:?} traceback={:?}", e.cause(), e.context(), e.traceback().unwrap());
        return Err(errmsg);
    }

    Ok(())
}

fn traverse_filesystem(root_dir: &str, userscript_filepath: &str) -> Result<(), String> {
    // (1) verify that all specified filepaths exist
    let userscript_path = path::Path::new(userscript_filepath);
    if !userscript_path.exists() {
        return Err(format!("userscript path not found '{}'", userscript_filepath));
    }

    // (2) initialize Python runtime
    let interp = rustpython::InterpreterConfig::new().init_stdlib().interpreter();
    if let Err(err) = interp.enter(|rvm| {
        let main_scope = rvm.new_scope_with_builtins();
        let source: String = fs::read_to_string(userscript_path).expect(&format!("python could not read user script '{}' file", userscript_path.display()));

        let compiled: Result<PyRef<PyCode>, CompileError> = rvm.compile(&source, Mode::Exec, userscript_path.display().to_string());
        match compiled {
            Ok(ref_code) => {
                rvm.run_code_obj(ref_code, main_scope.clone())?;
                println!("python source code was compiled");
                // TODO may I fetch a reference to python function 'rename' here?
            },
            Err(err) => {
                eprintln!("could not compile provided user script '{}'", userscript_path.display());
                eprintln!("error: {}", err);
                return Err(rvm.new_import_error(format!("failed to compile provided user script"), PyStr::from("userscript.py").into_ref(rvm.ctx.as_ref())));
            },
        }
        
        Ok(())
    }) {
        let e = err.clone();
        let errmsg = format!("cause={:?} context={:?} traceback={:?}", e.cause(), e.context(), e.traceback().unwrap());
        interp.finalize(Some(err.clone()));
        return Err(errmsg); 
    }

    // (3) traverse filesystem
    let root_path = path::PathBuf::from(root_dir);
    let walk = WalkDir::new(root_path).follow_links(true).sort_by_key(|a| a.file_name().to_owned()).into_iter().filter_map(|e| e.ok());
    for entry in walk {
        evaluate_filepath(&interp, entry.path(), userscript_path)?;
    }

    interp.finalize(None);

    Ok(())
}

fn main() {
    match traverse_filesystem(".", "userscript.py") {
        Err(err) => eprintln!("{}", err),
        Ok(_) => println!("done."),
    }
}
```

Now the question is:

* I currently compile the user script for every call of a filepath. This is inefficient.
* To change that, I need a reference to the userprovided function `rename`. However, the scope of `rename` is living within the `rvm`. And `rvm` is only available during the closure.
* Hence my question: Can I implement this functionality without rereading the user script every time?

**Comment by tajpulo** [comment]
Spelling it out, I think I found the solution myself:

* I don't know the design decisions behind it, but the Interpreter instance is completely irrelevant for my application and only the `rvm` is useful for me. I should not supply `interp: &vm::Interpreter` to `evaluate_filepath`, but `rvm: &VirtualMachine`.
* Once we have loaded the `VirtualMachine` instance, we should load the userscript once and then start traversing the file system. In the process we repeatedly call the user-provided function.
* We need to restructure our code to have one `vm::Interpreter` and `VirtualMachine` instance from the beginning. And almost the entire codebase is within the `if let Err(err) = interp.enter(|rvm| { … })` closure.
* In conclusion, we don't call an interpreter for each filepath, but each filepath for the one interpreter.

**Reply by youknowone** [reply]
You can enter interp multiple times.

Make first enter to return `fn1` and save it to reuse it later.
You can use it as many times as you need by reusing `fn1` variable later.

**Reply by tajpulo** [reply]
Indeed, you are right. I can just assign `fn1` to a variable and reuse it upon the next invocation of `enter`. I would never have guessed that something called `rustpython_vm::object::PyObjectRef` can be used across `enter` calls 😲

```rust
//! Application to run python userscript per filepath

use rustpython_vm as vm;
use rustpython;

use vm::builtins::{PyCode, PyStr};
use vm::compiler::CompileError;
use vm::{PyPayload, PyRef};
use rustpython_vm::compiler::core::Mode;

use walkdir::WalkDir;

use std::fs;
use std::path;

fn evaluate_filepath(interp: &vm::Interpreter, fn_ref: &rustpython_vm::object::PyObjectRef, filepath: &path::Path, userscript_filepath: &path::Path) -> Result<(), String> {
    let filepath_basename = match filepath.file_name() {
        Some(b) => b,
        None => panic!("filepath '{}' has no basename", filepath.display()),
    };

    if let Err(err) = interp.enter(|rvm| {
        let fn1_result = fn_ref.call((filepath_basename.to_str().unwrap().to_string(),), rvm)?;

        let ret_value: PyRef<PyStr> = match fn1_result.downcast() {
            Ok(string_return_value) => string_return_value,
            Err(_) => return Err(rvm.new_value_error(format!("function rename did not return a string"))),
        };
        println!("modified filepath = '{}'", ret_value);
        
        Ok(())
    }) {
        let e = err.clone();
        let errmsg = format!("cause={:?} context={:?} traceback={:?}", e.cause(), e.context(), e.traceback().unwrap());
        return Err(errmsg);
    }

    Ok(())
}

fn traverse_filesystem(root_dir: &str, userscript_filepath: &str) -> Result<(), String> {
    // (1) verify that all specified filepaths exist
    let userscript_path = path::Path::new(userscript_filepath);
    if !userscript_path.exists() {
        return Err(format!("userscript path not found '{}'", userscript_filepath));
    }

    // (2) initialize Python runtime
    let interp = rustpython::InterpreterConfig::new().init_stdlib().interpreter();
    let mut opt_rename_func: Option<rustpython_vm::object::PyObjectRef> = None;
    if let Err(err) = interp.enter(|rvm| {
        let main_scope = rvm.new_scope_with_builtins();
        let source: String = fs::read_to_string(userscript_path).expect(&format!("python could not read user script '{}' file", userscript_path.display()));

        let compiled: Result<PyRef<PyCode>, CompileError> = rvm.compile(&source, Mode::Exec, userscript_path.display().to_string());
        match compiled {
            Ok(ref_code) => {
                rvm.run_code_obj(ref_code, main_scope.clone())?;
                println!("python source code was compiled");

                let fn1 = main_scope.globals.get_item("rename", rvm)?;
                opt_rename_func = Some(fn1);
            },
            Err(err) => {
                eprintln!("could not compile provided user script '{}'", userscript_path.display());
                eprintln!("error: {}", err);
                return Err(rvm.new_import_error(format!("failed to compile provided user script"), PyStr::from("userscript.py").into_ref(rvm.ctx.as_ref())));
            },
        }
        
        Ok(())
    }) {
        let e = err.clone();
        let errmsg = format!("cause={:?} context={:?} traceback={:?}", e.cause(), e.context(), e.traceback().unwrap());
        interp.finalize(Some(err.clone()));
        return Err(errmsg); 
    }
    let rename_func: rustpython_vm::object::PyObjectRef = match opt_rename_func {
        Some(f_ref) => f_ref,
        None => panic!("user script function not found"),
    };

    // (3) traverse filesystem
    let root_path = path::PathBuf::from(root_dir);
    let walk = WalkDir::new(root_path).follow_links(true).sort_by_key(|a| a.file_name().to_owned()).into_iter().filter_map(|e| e.ok());
    for entry in walk {
        if entry.path() == path::Path::new(".") || entry.path() == path::Path::new("..") {
            continue
        }
        evaluate_filepath(&interp, &rename_func, entry.path(), userscript_path)?;
    }

    interp.finalize(None);

    Ok(())
}

fn main() {
    match traverse_filesystem(".", "userscript.py") {
        Err(err) => eprintln!("{}", err),
        Ok(_) => println!("done."),
    }
}
```

Thank you, @youknowone !
